src/main/java/hexgui/gui/HexGui.java

```diff
@@ -0,0 +1,2393 @@
+    }
+    if (ctrl) {
+      if (!shift) {
+        for (int i = 0; i < m_selected_cells.size(); i++) {
+          HexPoint p = m_selected_cells.get(i);
+          m_guiboard.setSelected(p, false);
+        }
+        m_selected_cells.clear();
+        m_guiboard.setSelected(point, true);
+        m_selected_cells.add(point);
+      } else {
+        int found_at = -1;
+        for (int i = 0; i < m_selected_cells.size() && found_at == -1; i++) {
+          if (m_selected_cells.get(i) == point) found_at = i;
+        }
+
+        if (found_at != -1) {
+          m_guiboard.setSelected(point, false);
+          m_selected_cells.remove(found_at);
+        } else {
+          m_guiboard.setSelected(point, true);
+          m_selected_cells.add(point);
+        }
+      }
+
+      m_guiboard.repaint();
+
+    } else {
+      String context = m_toolbar.getClickContext();
+      if (context.equals("play")) {
+        if (m_guiboard.getColor(point) == HexColor.EMPTY) {
+          humanMove(new Move(point, m_tomove));
+        } else if (isSwapAllowed()) {
+          humanMove(new Move(HexPoint.get("swap-pieces"), m_tomove));
+        }
+      } else if (context.equals("black")) {
+        if (m_guiboard.getColor(point) == HexColor.BLACK) {
+          addSetupMove(new Move(point, HexColor.EMPTY));
+        } else {
+          addSetupMove(new Move(point, HexColor.BLACK));
+        }
+      } else if (context.equals("white")) {
+        if (m_guiboard.getColor(point) == HexColor.WHITE) {
+          addSetupMove(new Move(point, HexColor.EMPTY));
+        } else {
+          addSetupMove(new Move(point, HexColor.WHITE));
+        }
+      }
+    }
+  }
+
+  public void fieldDoubleClicked(HexPoint point, boolean ctrl, boolean shift) {
+    fieldClicked(point, ctrl, shift);
+  }
+
+  public Vector<HexPoint> getSelectedCells() {
+    return m_selected_cells;
+  }
+
+  public HexColor getColorToMove() {
+    return m_tomove;
+  }
+
+  public void humanMove(Move move) {
+    play(move);
+    htpPlay(move);
+    htpShowboard();
+    if (!m_guiboard.isBoardFull() && m_preferences.getBoolean("auto-respond") && m_program != null)
+      htpGenMove(m_tomove);
+  }
+
+  /**
+   * Update the GUI to reflect the given move. Do this without any changes to the game tree or the
+   * HTP.
+   */
+  private void guiPlay(Move move) {
+    if (m_guiboard.isYBoard() && move.getPoint() == HexPoint.SWAP_PIECES) {
+      m_guiboard.swapColors();
+    } else if (move.getPoint() == HexPoint.SWAP_PIECES) {
+      m_guiboard.swapPieces();
+    } else {
+      m_guiboard.setColor(move.getPoint(), move.getColor());
+    }
+    m_guiboard.clearMarks();
+    markLastPlayedStone();
+  }
+
+  public boolean isSwapAllowed() {
+    // Count the number of pieces on the board.
+    int count = m_guiboard.numberOfPieces();
+    // Check whether the game tree allows swapping.
+    boolean isswap = m_current.isSwap();
+    return count == 1 && !isswap;
+  }
+
+  private void play(Move move) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.